storage/innobase/fil/fil0crypt.cc

```diff
@@ -3144,3 +3145,169 @@ fil_space_verify_crypt_checksum(
 
 	return(encrypted);
 }
+
+redo_log_key* redo_log_keys::load_latest_key() {
+	size_t   klen = 0;
+	size_t   klen2 = 0;
+	uint    version = 0;
+	char *key_type = NULL;
+	byte *rkey = NULL;
+	byte *rkey2 = NULL;
+
+	if (my_key_fetch(PERCONA_REDO_KEY_NAME, &key_type, NULL,
+		reinterpret_cast<void**>(&rkey), &klen) || rkey == NULL ||
+		strncmp(key_type, "AES", 3) != 0) {
+
+                /* There is no key yet, we'll try to generate one */
+		my_free(rkey);
+		return generate_and_store_new_key();
+	}
+
+
+	const bool err = (parse_system_key(rkey, klen, &version, &rkey2, &klen2) == reinterpret_cast<uchar*>(NullS));
+	if (err) {
+		my_free(rkey);
+		my_free(rkey2);
+		my_free(key_type);
+	}
+
+	ut_ad(klen2 ==  ENCRYPTION_KEY_LEN);
+
+	{
+		key_iterator it = m_keys.find(version);
+
+		if(it != m_keys.end() && it->second.present) {
+			return &it->second;
+		}
+	}
+
+	redo_log_key* rk = &m_keys[version];
+	rk->version = version;
+	rk->present = true;
+	memcpy(rk->key, rkey2, ENCRYPTION_KEY_LEN);
+
+	my_free(rkey);
+	my_free(rkey2);
+	my_free(key_type);
+
+	return rk;
+}
+
+redo_log_key* redo_log_keys::load_key_version(uint version) {
```

Comment:
It looks like it should be possible to extract common code from these two functions

Reply:
I don't think I could extract too many parts - a few lines at most, otherwise, there are small differences everywhere. And that would make the code less readable.